src/bytes/complete.rs

```diff
@@ -242,55 +242,29 @@ where
 {
   move |i: I| {
     let input = i;
+    let input_len = input.input_len();
 
-    match input.position(|c| !cond(c)) {
-      Some(idx) => {
-        if idx >= m {
-          if idx <= n {
-            let res: IResult<_, _, Error> = if let Ok(index) = input.slice_index(idx) {
-              Ok(input.take_split(index))
-            } else {
-              Err(Err::Error(Error::from_error_kind(
-                input,
-                ErrorKind::TakeWhileMN,
-              )))
-            };
-            res
-          } else {
-            let res: IResult<_, _, Error> = if let Ok(index) = input.slice_index(n) {
-              Ok(input.take_split(index))
-            } else {
-              Err(Err::Error(Error::from_error_kind(
-                input,
-                ErrorKind::TakeWhileMN,
-              )))
-            };
-            res
-          }
-        } else {
-          let e = ErrorKind::TakeWhileMN;
-          Err(Err::Error(Error::from_error_kind(input, e)))
-        }
+    for (i, item) in input.iter_elements().enumerate() {
+      if i > n {
+        break;
       }
-      None => {
-        let len = input.input_len();
-        if len >= n {
-          match input.slice_index(n) {
-            Ok(index) => Ok(input.take_split(index)),
-            Err(_needed) => Err(Err::Error(Error::from_error_kind(
-              input,
-              ErrorKind::TakeWhileMN,
-            ))),
+      if !cond(item) {
+        if i >= m {
+          if let Ok(index) = input.slice_index(min(i, n)) {
```

Comment:
```suggestion
          if let Ok(index) = input.slice_index(i) {
```
here we already know that i <=n

Reply:
maybe I removed `iter_indices` too early. it would have avoided going through the data twice here. I think we'll merge this and I'll see if I can simplify it